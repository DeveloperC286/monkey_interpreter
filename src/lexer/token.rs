#[derive(PartialEq, PartialOrd)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

#[derive(PartialEq, PartialOrd)]
#[allow(non_camel_case_types)]
pub enum TokenType {
    ILLEGAL,
    EOF,

    // Identifiers
    IDENTIFIER,
    INTEGER,

    // Operators
    ASSIGN,
    PLUS,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    OPENING_SQUARE_BRACKET,
    CLOSING_SQUARE_BRACKET,

    OPENING_ANGLE_BRACKET,
    CLOSING_ANGLE_BRACKET,

    //Keywords
    FUNCTION,
    LET,
}
