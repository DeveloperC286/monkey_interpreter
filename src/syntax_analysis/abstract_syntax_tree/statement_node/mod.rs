pub mod let_statement_node;

pub trait StatementNode {}
