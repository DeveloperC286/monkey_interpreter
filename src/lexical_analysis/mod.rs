use std::iter::{FromIterator, Peekable};
use std::str::Chars;

use crate::lexical_analysis::model::lexical_error::LexicalError;
use crate::lexical_analysis::model::token::Token;
use crate::lexical_analysis::utilities::*;

#[macro_use]
mod macros;

pub(crate) mod model;
mod utilities;

pub(crate) struct LexicalAnalysis<'a> {
    source_code: Peekable<Chars<'a>>,
}

impl<'a> LexicalAnalysis<'a> {
    pub(crate) fn from(code: &str) -> Result<Vec<Token>, LexicalError> {
        let mut lexical_analysis = LexicalAnalysis {
            source_code: code.chars().peekable(),
        };

        lexical_analysis.get_tokens()
    }

    fn get_tokens(&mut self) -> Result<Vec<Token>, LexicalError> {
        let mut tokens = Vec::new();

        loop {
            let token = self.get_next_token()?;
            let end_of_file = token == Token::EndOfFile;
            tokens.push(token);

            if end_of_file {
                break;
            }
        }

        Ok(tokens)
    }

    fn get_next_token(&mut self) -> Result<Token, LexicalError> {
        match self.get_next_character() {
            Some(character) => {
                debug!("Matching the character '{}'.", character);
                match character {
                    '!' => {
                        check_next_character!(self.source_code, '=', Ok(Token::NotEquals));
                        Ok(Token::Not)
                    }
                    '-' => Ok(Token::Minus),
                    '/' => Ok(Token::Divide),
                    '*' => Ok(Token::Multiply),
                    '>' => Ok(Token::GreaterThan),
                    '<' => Ok(Token::LesserThan),
                    '=' => {
                        check_next_character!(self.source_code, '=', Ok(Token::Equals));
                        Ok(Token::Assign)
                    }
                    '+' => Ok(Token::Plus),
                    '(' => Ok(Token::OpeningRoundBracket),
                    ')' => Ok(Token::ClosingRoundBracket),
                    '{' => Ok(Token::OpeningCurlyBracket),
                    '}' => Ok(Token::ClosingCurlyBracket),
                    ',' => Ok(Token::Comma),
                    ';' => Ok(Token::SemiColon),
                    _ => {
                        if is_valid_identifier_character(character) {
                            debug!("Parsing word from characters.");
                            return Ok(get_keyword_token(&self.get_word(character)));
                        }

                        if is_digit(character) {
                            debug!("Parsing integer from characters.");
                            return Ok(Token::Integer {
                                literal: self.get_integer(character)?,
                            });
                        }

                        Err(LexicalError::IllegalCharacter(character))
                    }
                }
            }
            None => Ok(Token::EndOfFile),
        }
    }

    fn get_integer_string(&mut self, character: char) -> String {
        parse_characters!(self.source_code, character, is_digit);
    }

    fn get_integer(&mut self, character: char) -> Result<i64, LexicalError> {
        let integer_string = self.get_integer_string(character);

        match integer_string.parse() {
            Ok(integer) => Ok(integer),
            Err(_) => Err(LexicalError::NonNumericIntegerString(integer_string)),
        }
    }

    fn get_word(&mut self, character: char) -> String {
        parse_characters!(self.source_code, character, is_valid_identifier_character);
    }

    fn get_next_character(&mut self) -> Option<char> {
        trace!("Getting next character.");
        let mut next_character = self.source_code.next();

        if let Some(character) = next_character {
            match character {
                ' ' | '\t' | '\n' | '\r' => {
                    trace!("Consuming formatting character.");
                    next_character = self.get_next_character();
                }
                _ => {}
            }
        }

        next_character
    }
}

#[cfg(test)]
mod tests;
