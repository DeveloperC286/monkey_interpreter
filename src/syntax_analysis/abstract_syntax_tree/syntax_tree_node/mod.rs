use crate::lexical_analysis::token::Token;

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub nodes: Vec<SyntaxTreeNode>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum SyntaxTreeNode {
    STATEMENT { statement: Statement },
    EXPRESSION { expression: Expression },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    LET {
        identifier_token: Token,
        expression: Box<Expression>,
    },
    RETURN {
        expression: Box<Expression>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    IDENTIFIER {
        identifier_token: Token,
    },
    INTEGER {
        integer_token: Token,
    },
    PREFIX {
        prefix_token: Token,
        right_hand: Box<Expression>,
    },
    INFIX {
        left_hand: Box<Expression>,
        operator_token: Token,
        right_hand: Box<Expression>,
    },
    BOOLEAN {
        boolean_token: Token,
    },
    IF {
        condition: Box<Expression>,
        consequence: Box<Block>,
        alternative: Box<Option<Block>>,
    },
    FUNCTION {
        parameters: Vec<Expression>,
        block: Box<Block>,
    },
    CALL {
        function: Box<Expression>,
        arguments: Vec<Expression>,
    },
}

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
#[allow(non_camel_case_types)]
pub enum ExpressionPrecedence {
    LOWEST,
    EQUALS,
    LESSER_OR_GREATER,
    PLUS,
    MULTIPLY,
    PREFIX,
    CALL,
}
