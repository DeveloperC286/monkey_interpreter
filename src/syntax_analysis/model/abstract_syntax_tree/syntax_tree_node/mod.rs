use crate::lexical_analysis::model::token::Token;

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub nodes: Vec<SyntaxTreeNode>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum SyntaxTreeNode {
    Statement { statement: Statement },
    Expression { expression: Expression },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    Let {
        identifier_token: Token,
        expression: Box<Expression>,
    },
    Return {
        expression: Box<Expression>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    Identifier {
        identifier_token: Token,
    },
    Integer {
        integer_token: Token,
    },
    Prefix {
        prefix_token: Token,
        right_hand: Box<Expression>,
    },
    Infix {
        left_hand: Box<Expression>,
        operator_token: Token,
        right_hand: Box<Expression>,
    },
    Boolean {
        boolean_token: Token,
    },
    If {
        condition: Box<Expression>,
        consequence: Box<Block>,
        alternative: Box<Option<Block>>,
    },
    Function {
        parameters: Vec<Expression>,
        block: Box<Block>,
    },
    Call {
        function: Box<Expression>,
        arguments: Vec<Expression>,
    },
}
