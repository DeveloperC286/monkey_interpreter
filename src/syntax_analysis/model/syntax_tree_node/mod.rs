use crate::lexical_analysis::model::token::Token;

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct Block {
    pub(crate) nodes: Vec<SyntaxTreeNode>,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum SyntaxTreeNode {
    Statement { statement: Statement },
    Expression { expression: Expression },
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum Statement {
    Let {
        identifier: String,
        expression: Box<Expression>,
    },
    Return {
        expression: Box<Expression>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum Expression {
    Identifier {
        identifier: String,
    },
    Integer {
        literal: i64,
    },
    String {
        literal: String,
    },
    Prefix {
        prefix: Token,
        right_hand: Box<Expression>,
    },
    Infix {
        left_hand: Box<Expression>,
        operator: Token,
        right_hand: Box<Expression>,
    },
    Boolean {
        literal: bool,
    },
    If {
        condition: Box<Expression>,
        consequence: Box<Block>,
        alternative: Box<Option<Block>>,
    },
    Function {
        parameters: Vec<String>,
        block: Box<Block>,
    },
    Call {
        function: Box<Expression>,
        arguments: Vec<Expression>,
    },
}
