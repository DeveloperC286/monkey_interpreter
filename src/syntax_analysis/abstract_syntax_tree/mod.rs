use crate::syntax_analysis::abstract_syntax_tree::syntax_tree_node::SyntaxTreeNode;

pub mod syntax_tree_node;

#[derive(Debug)]
pub struct AbstractSyntaxTree {
    pub abstract_syntax_tree: Vec<SyntaxTreeNode>,
    pub syntax_parsing_errors: Vec<String>,
}
