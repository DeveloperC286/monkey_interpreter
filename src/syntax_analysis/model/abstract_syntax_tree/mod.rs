use crate::syntax_analysis::model::abstract_syntax_tree::syntax_tree_node::SyntaxTreeNode;

pub(crate) mod syntax_tree_node;

#[derive(Debug)]
pub(crate) struct AbstractSyntaxTree {
    pub(crate) abstract_syntax_tree: Vec<SyntaxTreeNode>,
    pub(crate) syntax_parsing_errors: Vec<String>,
}
