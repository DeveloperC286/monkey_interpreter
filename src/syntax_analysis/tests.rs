use super::super::lexical_analysis::token::{Token, TokenType};
use super::abstract_syntax_tree::syntax_tree_node::SyntaxTreeNode;
use super::abstract_syntax_tree::AbstractSyntaxTree;

use super::*;
use rstest::rstest;

#[rstest(
    tokenized_let_statement,
    expected_abstract_syntax_tree,
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        AbstractSyntaxTree{ program: vec![
            SyntaxTreeNode::LetStatement {
                let_token: Token{token_type: TokenType::LET, literal: "let".to_string()},
                identifier_token: Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            },
        ]},
    ),
    case(
        vec![
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "5".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::LET, literal: "let".to_string()},
            Token{token_type: TokenType::IDENTIFIER, literal: "z".to_string()},
            Token{token_type: TokenType::ASSIGN, literal: "=".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "7".to_string()},
            Token{token_type: TokenType::PLUS, literal: "+".to_string()},
            Token{token_type: TokenType::INTEGER, literal: "10".to_string()},
            Token{token_type: TokenType::SEMI_COLON, literal: ";".to_string()},
            Token{token_type: TokenType::EOF, literal: "".to_string()},
        ],
        AbstractSyntaxTree{ program: vec![
            SyntaxTreeNode::LetStatement {
                let_token: Token{token_type: TokenType::LET, literal: "let".to_string()},
                identifier_token: Token{token_type: TokenType::IDENTIFIER, literal: "x".to_string()},
            },
            SyntaxTreeNode::LetStatement {
                let_token: Token{token_type: TokenType::LET, literal: "let".to_string()},
                identifier_token: Token{token_type: TokenType::IDENTIFIER, literal: "z".to_string()},
            },
        ]},
    ),


)]
fn test_syntax_analysis_for_let_statements(
    tokenized_let_statement: Vec<Token>,
    expected_abstract_syntax_tree: AbstractSyntaxTree,
) {
    //given
    let mut syntax_analysis = SyntaxAnalysis::new(tokenized_let_statement);

    //when
    let returned_abstract_syntax_tree = syntax_analysis.parse();

    //then
    assert_abstract_syntax_tree_equal(expected_abstract_syntax_tree, returned_abstract_syntax_tree);
}

fn assert_abstract_syntax_tree_equal(
    expected_abstract_syntax_tree: AbstractSyntaxTree,
    returned_abstract_syntax_tree: AbstractSyntaxTree,
) {
    assert_eq!(
        expected_abstract_syntax_tree.program.len(),
        returned_abstract_syntax_tree.program.len()
    );
    assert_eq!(
        expected_abstract_syntax_tree.program,
        returned_abstract_syntax_tree.program
    );
}
