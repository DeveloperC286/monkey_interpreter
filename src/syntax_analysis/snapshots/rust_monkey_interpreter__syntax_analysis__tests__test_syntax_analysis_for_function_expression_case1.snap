---
source: src/syntax_analysis/tests.rs
expression: returned_abstract_syntax_tree
---
{
  "abstract_syntax_tree": [
    {
      "EXPRESSION": {
        "expression": {
          "FUNCTION": {
            "parameters": [
              {
                "IDENTIFIER": {
                  "identifier_token": {
                    "token_type": "IDENTIFIER",
                    "literal": "x"
                  }
                }
              },
              {
                "IDENTIFIER": {
                  "identifier_token": {
                    "token_type": "IDENTIFIER",
                    "literal": "y"
                  }
                }
              }
            ],
            "block": {
              "BLOCK": {
                "blocks": [
                  {
                    "EXPRESSION": {
                      "expression": {
                        "INFIX": {
                          "left_hand_expression": {
                            "IDENTIFIER": {
                              "identifier_token": {
                                "token_type": "IDENTIFIER",
                                "literal": "x"
                              }
                            }
                          },
                          "operator_token": {
                            "token_type": "PLUS",
                            "literal": "+"
                          },
                          "right_hand_expression": {
                            "IDENTIFIER": {
                              "identifier_token": {
                                "token_type": "IDENTIFIER",
                                "literal": "y"
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                ]
              }
            }
          }
        }
      }
    }
  ],
  "syntax_parsing_errors": []
}
